File: l33tcode/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
import heapq
from typing import List


class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        heap = []

        count = 0

        for node in lists:
            if node is not None:
                count += 1
                heapq.heappush(heap, (node.val, count, node))

        root = ListNode(None)

        node = root

        while True:
            count += 1
            if not heap:
                break

            min_node_val, c, min_node = heapq.heappop(heap)

            node.next = min_node
            node = node.next

            if min_node.next is not None:
                heapq.heappush(heap, (min_node.next.val, count, min_node.next))


# Definition for singly-linked list.
class ListNodeComparable:
    def __init__(self, node: ListNode):
        self.val = node.val
        self.next = node.next
        self.node = node

    def __eq__(self, other: object) -> bool:
        return self.val == other.val

    def __lt__(self, other: object) -> bool:
        return self.val < other.val


class Solution2:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        heap: List[ListNodeComparable] = []

        head = ListNode()

        for node in lists:
            if node:
                heapq.heappush(heap, ListNodeComparable(node))

        node = head
        while heap:
            cur_node = heapq.heappop(heap)

            if cur_node.next:
                heapq.heappush(heap, ListNodeComparable(cur_node.next))

            node.next = cur_node.node
            node = node.next

        return head.next

File: l33tcode_rs/src/merge_k_sorted_lists.rs
#[allow(dead_code)]
struct Solution;

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

use std::{cmp::Reverse, collections::BinaryHeap};

impl Ord for ListNode {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.val.cmp(&other.val)
    }
}

impl PartialOrd for ListNode {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Solution {
    #[allow(dead_code)]
    pub fn merge_k_lists(lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>> {
        let mut heap = BinaryHeap::new();

        for node in lists.into_iter().flatten() {
            heap.push(Reverse(node));
        }

        if heap.is_empty() {
            return None;
        }

        let mut dummy = ListNode::new(-1);

        let mut current = &mut dummy;

        while let Some(Reverse(mut node)) = heap.pop() {
            if let Some(next_node) = node.next.take() {
                heap.push(Reverse(next_node));
            }

            current.next = Some(node);
            current = current.next.as_mut().unwrap();
        }

        dummy.next
    }
}
